pkg/controller/constraint/constraint_controller.go

```diff
@@ -153,12 +203,58 @@ func (r *ReconcileConstraint) Reconcile(request reconcile.Request) (reconcile.Re
 			if err := r.Update(context.Background(), instance); err != nil {
 				return reconcile.Result{Requeue: true}, nil
 			}
+			// removing constraint entry from cache
+			r.deleteConstraintKey(constraintKey)
 		}
 	}
-
+	reportMetrics = true
```

Comment:
nit: reportMetrics should only be set to true near addConstraintKey/deleteConstraintKey as those are the only commands that would mutate the metrics cache

Reply:
wouldn't adding it here would cover both non-error add and delete scenarios?

Reply:
Adding it here guarantees it runs every time the function doesn't return early, but unless there has been an add or delete, there would be no change to report, which means the metrics are recalculated (and locked) when there is no change in output.

Reply:
Adding it here guarantees it runs every time the function doesn't return early, but unless there has been an add or delete, there would be no change to report, which means the metrics are recalculated (and locked) when there is no change in output.

Reply:
Adding it here guarantees it runs every time the function doesn't return early, but unless there has been an add or delete, there would be no change to report, which means the metrics are recalculated (and locked) when there is no change in output.

Reply:
Actually, it looks like it wouldn't run more than it needed to as all non-mutate cases would have exited by then.

Coding-style wise it's still better to keep the `reportMetrics = true` adjacent to the mutate statements to guard against the time when the code changes and, say, delete exits early. It also prevents needing to look at the entire function logic to satisfy the "are we running this too much" question I gave the wrong answer to above.

Reply:
One last thing I remembered. Each reconciler is single-threaded, but there are multiple reconcilers. We need toake constraintsCacheMux a global variable so its shared across threads.

Reply:
thanks, good call on shared mutex. added/moved into util package with shared mutex, and moved reportMetrics